### DISCUSSION TITLE
Make it possible to Implement custom Inline usable in TextBlock

### DISCUSSION THREAD

**Discussion by techseat** [AvaloniaUI/Avalonia#13786] [Ideas]
As currently it is possible to inheriate from `DrawableTextRun`, Suppose make inheriating `Avalonia.Controls.Documents.Inline` possible to let developers create a new custom inline (form example `IconInline`).

Currently, `internal` modifier of methods `internal override void BuildTextRun(IList<TextRun> textRuns)` and `internal override void AppendText(StringBuilder stringBuilder)` prevents us from inheriating from `Avalonia.Controls.Documents.Inline`.

I am not sure this is the best way to achieve the purpose of making developing of `TextBlock` capabilities possible (to cover some needs in lack of `FlowDocument`), But I think it is a good consideration to let developing Text or Document capabilities without need to reinvent the weel (say without need to use basic things such as `TextFormatter`, etc). Is there a standard way already to achieve similar functionality that I was not aware of?

**Comment by timunie** [comment]
You can use SelectableTextBlock which already support it given you use 11.0.

**Reply by techseat** [reply]
OK. I tested by last nightly build, and it seems there is `InlineUIContainer` now. Possibly, user can simulate some use cases, But it seems this will not satisfy all the needs. As I understand, `InlineUIContainer.Child` will get part in Visual Tree and this is not suitable for some use cases.

Why cannot inheriate from `Avalonia.Controls.Documents.Inline` to develop vary inlines. For example, someone may need to develop a document-based app with its own complex drawing document elements. Can you give me some more details or guides about how can I achieve this?
Can I get to know the logic behind of preventing developer from inheriate `Avalonia.Controls.Documents.Inline` and extending `TextBlock` with custom app-specific features?

**Reply by cfnfy** [reply]
Do you have any further details of your requirements? 

Whilst there are currently no rich text editing controls in Avalonia, we are currently implementing our own custom solution. It has limitations in terms of inline controls right now, but it does allow us to type & paste text runs of different styles. 

We've also explored Chromium embedded QuillJS and SyncFusion WPF RichTextBoxAdv wrapped with XPF. Both would be great options in some cases, but unfortunately neither completely covered our requirements.

I'd be happy to help point you in the right direction. We've had to learn a lot over the past months, and if you do want to custom build there is a lot you can leverage from the existing code base. 

Depending on your budget, you may also look into speaking with the Avalonia team. I can't speak highly enough about how great they were helping us to where we are now.

**Comment by Gillibald** [comment]
The inlines API isn't designed to be extendable. You can always copy the SelectableTextBlock implementation and add you own version of inlines. In the end all this uses the TextFormatter.

In a future version of Avalonia the TextBlock will use the TextFormatter directly so you will be able to customize formatting via custom text runs.

**Reply by techseat** [reply]
@Gillibald, As
> The inlines API isn't designed to be extendable

And there is `DrawableTextRun` already exists, I suggest a new Inline to support `Drawing` to let developers use Images (and other drawings) in text flow. It seems this will not needs a large effort but brings valuable capabilities to `TextBlock`.

**Reply by Gillibald** [reply]
I am open for accepting such a feature. Feel free to submit a PR.
